src/intrinsics/ecma262/Object.js

```diff
@@ -101,36 +136,69 @@ export default function(realm: Realm): NativeFunctionValue {
-            }
+        handleSnapshot(frm, frm_was_partial);
+
+        // c. Repeat for each element nextKey of keys in List order,
+        invariant(frm, "from required");
+        invariant(keys, "keys required");
+        copyKeys(keys, frm, to);
+      };
+
+      const tryAndApplySourceOrRecover = nextSource => {
+        let effects;
+        let savedSuppressDiagnostics = realm.suppressDiagnostics;
+        try {
+          realm.suppressDiagnostics = true;
+          effects = realm.evaluateForEffects(
+            () => {
+              applySource(nextSource);
+              return realm.intrinsics.undefined;
+            },
+            undefined,
+            "tryAndApplySourceOrRecover"
+          );
+        } catch (e) {
+          // ignore null or undefined
+          if (nextSource === realm.intrinsics.null || nextSource === realm.intrinsics.undefined) {
+            return;
+          }
+          let frm = To.ToObject(realm, nextSource);
+
+          if (e instanceof FatalError && to.isSimpleObject()) {
+            let frm_was_partial = frm.isPartialObject();
 
             if (frm_was_partial) {
-              if (frm instanceof AbstractObjectValue && frm.kind === "explicit conversion to object") {
-                // Make it implicit again since it is getting delayed into an Object.assign call.
-                delayedSources.push(frm.args[0]);
-              } else {
-                frmSnapshot = frm.getSnapshot();
-                frm.temporalAlias = frmSnapshot;
-                frm.makePartial();
-                delayedSources.push(frmSnapshot);
-              }
+              to_must_be_partial = true;
             }
+            handleSnapshot(frm, frm_was_partial);
+            return;
           }
+          throw e;
+        } finally {
+          realm.suppressDiagnostics = savedSuppressDiagnostics;
         }
+        // Note that the effects of (non joining) abrupt branches are not included
+        // in effects, but are tracked separately inside completion.
+        realm.applyEffects(effects);
+        let completion = effects.result;
+        if (completion instanceof PossiblyNormalCompletion) {
+          // in this case one of the branches may complete abruptly, which means that
+          // not all control flow branches join into one flow at this point.
+          // Consequently we have to continue tracking changes until the point where
+          // all the branches come together into one.
+          completion = realm.composeWithSavedCompletion(completion);
+        }
+        // return or throw completion
+        if (completion instanceof AbruptCompletion) throw completion;
+      };
 
-        // c. Repeat for each element nextKey of keys in List order,
-        invariant(frm, "from required");
-        invariant(keys, "keys required");
-        copyKeys(keys, frm, to);
+      // 4. For each element nextSource of sources, in ascending index order,
```

Comment:
It's unfortunate that we've lost some of the spec text structure here by reordering it.

Reply:
I'll look to see if I can clean up the text structure some more.